### PR TITLE
Capture instance name

### PR DESCRIPTION
**Pull request by zackmay**
added a safety net if folks forget to add varnish instance name.

### CODE REVIEW

**Comment by gquintard** [comment]
@hermunn , I could use your opinion on this.
On the one hand, I like how straightforward the code is, it makes a low-effort, high probability attempt at fixing the initial command.
On the other hand, I feel the amount of magic and lack of transparency disturbing.

My knee-jerk reaction would be to run `varnishadm ping` (possibly with the user-provided `-n` argument) and warn if it fails ("is varnish running, do you have the right permissions, etc."), but that's actually less useful than the current proposal

**Comment by hermunn** [comment]
I think @gquintard has a good point. I think maybe a combination of the ideas would be the following:

* If no `-n` argument is given, you check if `varnishadm ping` (with the `-n` argument) gives a good answer. If it does not, you use this PR's idea to suggest what the users should do: _You need to / should consider to specify the instance using `-n`, and the following values seem to be reasonable:_ and then a list of the observed `varnishd` instances which have a `-n` argument. If there are no `varnishd` running, you finish this script, for the sake of other information gathering.
* If there is a `-n` argument, but `varnishadm ping -n` does not work, report that and suggest other `-n` values that might be better, but run the rest of the script.

In short, I think that the idea of figuring out the right `-n` is very good, but maybe it can make even better than what is in the gather. I would also be fine with merging this or something very similar to what we have, and expand on it later.

**Comment by gquintard** [comment]
@zackmay, bump

**Comment by zackmay** [comment]
I believe I addressed the concerns.  I've tested it on my test box, I am not sure on one piece.  Which is.  If someone provide the incorrect name, should we allow that which would result in the incorrect name represented in the file output.  I went ahead and fixed that to use the valid name, but I think that is best.  Can you guy please check my work?
Thanks

File: varnishgather [hunk 1]
```diff
@@ -169,12 +169,20 @@ vadmin() {
 	fi
 }
 
+get_varnishd_instances() {
+    ps -eo cmd | grep '[v]arnishd' | grep -oP '(?<=-n )\S+' | sort | uniq
```
Comment (gquintard) [inline varnishgather:173]:
```suggestion
    ps -eo cmd | grep '[v]arnishd' | grep -oP '(?<=-n )\S+' | sort -u
```

File: varnishgather [hunk 2]
```diff
@@ -393,6 +401,10 @@ _EOF_
 	exit $1
 }
 
+suggest_n_values() {
```
Comment (gquintard) [inline varnishgather:404]: this looks like an unnecessary level of indirection? If you want to use `get_varnishd_instances`, just use `get_varnishd_instances`

File: varnishgather [hunk 3]
```diff
@@ -428,6 +440,46 @@ done
 
 shift $((OPTIND - 1))
 
+if [ -z "$NAME" ]; then
+    VARNISHD_CMD=$(ps -o cmd= -p $(pgrep varnishd))
```
Comment (gquintard) [inline varnishgather:444]: this would generate multiple values when multiple `varnishd` processes are running, wouldn't it?

also, you wrote `get_varnishd_instances`, it feels like a shame to not use it?

File: varnishgather [hunk 4]
```diff
@@ -428,6 +440,46 @@ done
...
+            else
+                NAME=""
+            fi
+        else
+            NAME=""
+        fi
+    fi
+fi
+
+if [ -n "$NAME" ] && [ -z "$RELDIR" ]; then
```
Comment (gquintard) [inline varnishgather:476]: this block feels overkill for me, I have the impression that you just need to do
```
RELDIR="$RELDIR-$NEWLYFIGUREDOUTNAME"
ID="$ID-$NEWLYFIGUREDOUTNAME"
```
if `NAME` wasn't specified
if

File: varnishgather [hunk 5]
```diff
@@ -428,6 +440,46 @@ done
...
 shift $((OPTIND - 1))
 
+if [ -z "$NAME" ]; then
+    VARNISHD_CMD=$(ps -o cmd= -p $(pgrep varnishd))
+    INSTANCE_NAME=$(echo "$VARNISHD_CMD" | grep -oP '(?<=-n )\S+')
+
+    if [ -n "$INSTANCE_NAME" ]; then
+        NAME="-n $INSTANCE_NAME"
+    else
+        INSTANCE_NAME=$(suggest_n_values)
```
Comment (gquintard) [inline varnishgather:450]: why not start with that?

Comment (zackmay) [reply varnishgather:450]: Covered in the previous two updates I hope.

File: varnishgather (hunk 4 above)
Comment (zackmay) [reply varnishgather:476]: I moved this into the block if block 439-452 after thinking about it I just reused INSTANCE_NAME set from get_varnishd_instance

File: varnishgather (hunk 3 above)
Comment (zackmay) [reply varnishgather:444]: I changed this entire block to tighten up sloppy work on my part.

File: varnishgather [hunk 6]
```diff
@@ -428,6 +436,21 @@ done
...
+    INSTANCE_NAME=$(get_varnishd_instance)
+    [ -n "$INSTANCE_NAME" ] && NAME="-n $INSTANCE_NAME"
+else
+    INSTANCE_NAME="${NAME#-n }"
+fi
+
+if [ -n "$NAME" ] && ! validate_instance; then
+    INSTANCE_NAME=$(get_varnishd_instance)
+    [ -n "$INSTANCE_NAME" ] && NAME="-n $INSTANCE_NAME" || NAME=""
+fi
```
Comment (gquintard) [inline varnishgather:449]: Pål wrote:

> If there is a -n argument, but varnishadm ping -n does not work, report that and suggest other -n values that might be better, but run the rest of the script.

I don't think this code is doing that

File: varnishgather [hunk 7]
```diff
@@ -428,6 +436,21 @@ done
...
+
+ID="$(cat /etc/hostname)-$(date +'%Y%m%d-%H%M%S')${INSTANCE_NAME:+-$(n_opt "$INSTANCE_NAME")}"
```
Comment (gquintard) [inline varnishgather:451]: isn't that just?
```
ID="$(cat /etc/hostname)-$(date +'%Y%m%d-%H%M%S')$(n_opt "$INSTANCE_NAME")"
```
since `n_opt` will be a noop if `INSTANCE_NAME` is empty?

File: varnishgather (hunk 6 above)
Comment (zackmay) [reply varnishgather:449]: you are right, I did not go back to your original comments I focused on fixing my laziness before updating.  I'll work make sure to add a varnishadm ping check in here, my bad

Comment (zackmay) [reply varnishgather:449]: after re-reading your original statement.  I completely skipped over what you wanted.  I focused on just ensuring the name was included in the script execution with or without the client adding it, or even if they add the wrong name it picks the proper name.  A new day and a revisit to your comment Pal has me feeling you would rather have the following.

1. script executed, check provided -n value (if correct it goes on about it's marry way)
2. script executed, no -n provided but there is a name, so we spit out to the client they should rerun the script with -n $propername (exit script)
3.. script executed with -n and wrong name, return the potential proper names we gather and present that to the client with the NAME's we found with echo statements to run the script properly but then exit script.

I went about this task thinking I should just do the -n work for them.  There is a potential for compromise here.  I could go do the -n capture and then also, print out to the client that we had to do this because they failed to provide the information.  Even warning we chose the first -n process we found, therefore if they meant X/Y/Z then they should re-run the script with the proper name?

Comment (gquintard) [reply varnishgather:449]: to be completely exhaustive, I think this is the proper logic:
- `-n` given
  - `varnishadm ping` works
    - proceed as usual
  - `varnishadm ping` doesn't work
    - warn that the name may be wrong, print out a list of candidates if any, then proceed
- no `-n`
  - `varnishadm ping` works
    - proceed as usual
  - if there's exactly one candidate
    - warn that we defaulted to that auto-selected name, proceed if `varnishadm ping` works, error out otherwise
  - there is 0 candidate
    - warn that `varnishd` may not be running
  - there are more than 1 candidate
    - list the candidates, error out

@zackmay , @hermunn , how does it look?

File: varnishgather (hunk 7 above)
Comment (zackmay) [reply varnishgather:451]: The reason I have it the way it is.  When using just ID="$(cat /etc/hostname)-$(date +'%Y%m%d-%H%M%S')$(n_opt "$INSTANCE_NAME")"
the results from a named instance output look like below.
Please submit the file: varnishgather.varnish-wrk-20241101-184253varnish1.tar.gz

**Comment by zackmay** [comment]
I've pushed the latest adoption to the script to address your concerns.  
1. -n + actual name proceed and output file
2. -n + improper name + single instance = warning + corrected name
3. -n + multiple named instances = produced list and exit script
4. no -n captures the unnamed varnishd instance and produces resulting output file. 

Tests I ran
3 varnish instances, 2 named, one unnamed, all three running
1. tested script with no -n captured the unnamed instance and produced the tar.gz file that instance
2. tested script with -n wrong name, produced list of available names
3. 1 named and one unnamed instance running, reran test 1, then ran with -n wrong name, received warning, and output was the properly named file

To note, there was no varnishping processing in the original script, which is why I proceeded in this direction of capturing the unnamed instance as primary if -n was not provided.  If -n, new logic is processed. 

I hope that addresses your concerns.  I also commented above on your n_opt usage.

File: varnishgather (hunk 7 above)
Comment (gquintard) [reply varnishgather]: what I'm saying is that `$(n_opt "$INSTANCE_NAME")"` is a shorter equivalent to `${INSTANCE_NAME:+-$(n_opt "$INSTANCE_NAME")}"`

File: varnishgather [hunk 8]
```diff
@@ -428,6 +436,40 @@ done
 
 shift $((OPTIND - 1))
 
+INSTANCE_NAMES=$(get_varnishd_instance)
+if [ -n "$NAME" ]; then
+    INSTANCE_NAME="${NAME#-n }"
+   
+    if ! echo "$INSTANCE_NAMES" | grep -qx "$INSTANCE_NAME"; then
```
Comment (gquintard) [inline varnishgather:443]: this is sketchy because an unnamed instance can still be pinged using `-n`, and you can use an absolute `-n` to ping a relatively named instance. As long as the workdir can be fine it work.

However, I don't foresee users doing twisted things like that, this is fine
